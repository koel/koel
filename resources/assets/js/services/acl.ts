import { http } from '@/services/http'
import { cache } from '@/services/cache'

type ResourceType = 'album' | 'artist' | 'radio-station' | 'user'
type Action = 'edit' | 'delete'

export const acl = {
  checkResourcePermission: async (type: ResourceType, id: string | number, action: Action) => {
    return await cache.remember(['permission', type, id, action], async () => {
      const { allowed } = await http.silently.get<{ allowed: boolean }>(`acl/permissions/${type}/${id}/${action}`)

      return allowed
    })
  },

  fetchAssignableRoles: async () => {
    return await cache.remember(['assignable-roles'], async () => {
      const { roles } = await http.get<{
        roles: Array<{ id: Role, label: string, description: string }>
      }>('acl/assignable-roles')

      return roles
    })
  },
}
