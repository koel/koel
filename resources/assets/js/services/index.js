export * from "./info";
export * from "./download";
export * from "./http";
export * from "./ls";
export * from "./playback";
export * from "./youtube";
export * from "./socket";
