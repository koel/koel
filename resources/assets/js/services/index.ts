export * from './http'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
export * from './mediaInfoService'
export * from './cache'
export * from './socketListener'
export * from './volumeManager'
