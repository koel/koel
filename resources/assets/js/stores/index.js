export * from "./album";
export * from "./artist";
export * from "./equalizer";
export * from "./favorite";
export * from "./playlist";
export * from "./preference";
export * from "./queue";
export * from "./setting";
export * from "./shared";
export * from "./song";
export * from "./user";
