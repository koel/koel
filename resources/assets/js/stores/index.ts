export * from './albumStore'
export * from './artistStore'
export * from './commonStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './overviewStore'
export * from './playlistStore'
export * from './playlistFolderStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './searchStore'
export * from './settingStore'
export * from './songStore'
export * from './themeStore'
export * from './userStore'
export * from './genreStore'
export * from './visualizerStore'
