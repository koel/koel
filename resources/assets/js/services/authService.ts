import { localStorageService } from '@/services'

const API_TOKEN_STORAGE_KEY = 'api-token'
const AUDIO_TOKEN_STORAGE_KEY = 'audio-token'

export const authService = {
  getApiToken: () => localStorageService.get(API_TOKEN_STORAGE_KEY),

  hasApiToken () {
    return Boolean(this.getApiToken())
  },

  setApiToken: (token: string) => localStorageService.set(API_TOKEN_STORAGE_KEY, token),

  destroy: () => {
    localStorageService.remove(API_TOKEN_STORAGE_KEY)
    localStorageService.remove(AUDIO_TOKEN_STORAGE_KEY)
  },

  setAudioToken: (token: string) => localStorageService.set(AUDIO_TOKEN_STORAGE_KEY, token),

  getAudioToken: () => {
    // for backward compatibility, we first try to get the audio token, and fall back to the (full-privileged) API token
    return localStorageService.get(AUDIO_TOKEN_STORAGE_KEY) || localStorageService.get(API_TOKEN_STORAGE_KEY)
  }
}
