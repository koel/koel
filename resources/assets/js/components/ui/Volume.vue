<template>
  <span id="volume" class="volume" :class="level">
    <span
      v-show="level === 'muted'"
      v-koel-tooltip.top
      role="button"
      tabindex="0"
      title="Unmute"
      @click="unmute"
    >
      <icon :icon="faVolumeMute" fixed-width/>
    </span>

    <span
      v-show="level !== 'muted'"
      v-koel-tooltip.top
      role="button"
      tabindex="0"
      title="Mute"
      @click="mute"
    >
      <icon :icon="level === 'discreet' ? faVolumeLow : faVolumeHigh" fixed-width/>
    </span>

    <input
      id="volumeInput"
      class="plyr__volume"
      max="10"
      role="slider"
      step="0.1"
      title="Volume"
      type="range"
      @change="broadcastVolume"
      @input="setVolume"
    >
  </span>
</template>

<script lang="ts" setup>
import { faVolumeHigh, faVolumeLow, faVolumeMute } from '@fortawesome/free-solid-svg-icons'
import { ref } from 'vue'
import { playbackService, socketService } from '@/services'
import { preferenceStore as preferences } from '@/stores'
import { eventBus } from '@/utils'

const level = ref<'muted' | 'discreet' | 'loud'>()

const mute = () => {
  playbackService.mute()
  level.value = 'muted'
}

const unmute = () => {
  playbackService.unmute()
  level.value = preferences.volume < 3 ? 'discreet' : 'loud'
}

const setVolume = (e: InputEvent) => {
  const volume = parseFloat((e.target as HTMLInputElement).value)
  playbackService.setVolume(volume)
  setLevel(volume)
}

const setLevel = (volume: number) => (level.value = volume === 0 ? 'muted' : volume < 3 ? 'discreet' : 'loud')

/**
 * Broadcast the volume changed event to remote controller.
 */
const broadcastVolume = (e: InputEvent) => {
  socketService.broadcast('SOCKET_VOLUME_CHANGED', parseFloat((e.target as HTMLInputElement).value))
}

eventBus.on('KOEL_READY', () => setLevel(preferences.volume))
</script>

<style lang="scss">
#volume {
  position: relative;
  display: flex;
  align-items: center;

  [type=range] {
    margin: 0 0 0 8px;
    width: 120px;
    height: 4px;
    border-radius: 4px;
    position: relative;

    // increase click area
    &::before {
      position: absolute;
      content: ' ';
      left: 0;
      right: 0;
      top: -12px;
      bottom: -12px;
    }

    &::-webkit-slider-thumb {
      background: var(--color-text-secondary);
    }

    &:hover {
      &::-webkit-slider-thumb {
        background: var(--color-text-primary);
      }
    }
  }

  &.muted {
    [type=range] {
      &::-webkit-slider-thumb {
        background: transparent;
        box-shadow: none;
      }
    }
  }

  @media only screen and (max-width: 768px) {
    display: none !important;
  }
}
</style>
