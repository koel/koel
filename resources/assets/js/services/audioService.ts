import { equalizerStore } from '@/stores'
import { frequencies } from '@/config'
import { dbToGain } from '@/utils'

interface Band {
  label: string
  filter: BiquadFilterNode
  db: number
}

export const audioService = {
  unlocked: false,

  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,
  preampGainNode: null as unknown as GainNode,
  analyzer: null as unknown as AnalyserNode,

  bands: [] as Band[],

  init (mediaElement: HTMLMediaElement) {
    this.element = mediaElement

    this.context = new AudioContext()
    this.preampGainNode = this.context.createGain()
    this.source = this.context.createMediaElementSource(this.element)
    this.analyzer = this.context.createAnalyser()

    this.source.connect(this.preampGainNode)

    const config = equalizerStore.getConfig()

    this.changePreampGain(config.preamp)

    let prevFilter: BiquadFilterNode

    // Create 10 bands with the frequencies similar to those of Winamp and connect them together.
    frequencies.forEach((frequency, i) => {
      const filter = this.context.createBiquadFilter()

      if (i === 0) {
        filter.type = 'lowshelf'
      } else if (i === frequencies.length - 1) {
        filter.type = 'highshelf'
      } else {
        filter.type = 'peaking'
      }

      filter.Q.setTargetAtTime(1, this.context.currentTime, 0.01)
      filter.frequency.setTargetAtTime(frequency, this.context.currentTime, 0.01)
      filter.gain.value = dbToGain(config.gains[i])

      prevFilter ? prevFilter.connect(filter) : this.preampGainNode.connect(filter)
      prevFilter = filter

      this.bands.push({
        filter,
        label: String(frequency).replace('000', 'K'),
        db: config.gains[i]
      })
    })

    prevFilter!.connect(this.analyzer)

    // connect the analyzer node last, so that changes to the equalizer affect the visualizer as well
    this.analyzer.connect(this.context.destination)

    this.unlockAudioContext()
  },

  changePreampGain (db: number) {
    this.preampGainNode.gain.value = dbToGain(db)
  },

  changeFilterGain (node: BiquadFilterNode, db: number) {
    this.bands.find(band => band.filter === node)!.db = db
    node.gain.value = dbToGain(db)
  },

  /**
   * Attempt to unlock the audio context on mobile devices by creating and playing a silent buffer upon the
   * first user interaction.
   */
  unlockAudioContext () {
    ['touchend', 'touchstart', 'click'].forEach(event => {
      document.addEventListener(event, () => {
        if (this.unlocked) return

        const source = this.context.createBufferSource()
        source.buffer = this.context.createBuffer(1, 1, 22050)
        source.connect(this.context.destination)
        source.start(0)

        this.unlocked = true
      }, {
        once: true
      })
    })
  }
}
