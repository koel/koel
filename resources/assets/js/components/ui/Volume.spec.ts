import { expect, it } from 'vitest'
import UnitTestCase from '@/__tests__/UnitTestCase'
import { fireEvent, screen } from '@testing-library/vue'
import { socketService, volumeManager } from '@/services'
import { preferenceStore } from '@/stores'
import Volume from './Volume.vue'

new class extends UnitTestCase {
  protected beforeEach (cb?: Closure) {
    super.beforeEach(() => {
      preferenceStore.volume = 5
      volumeManager.init(document.createElement('input'))
    })
  }

  protected test () {
    it('mutes and unmutes', async () => {
      const { html } = this.render(Volume)
      expect(html()).toMatchSnapshot()
      expect(volumeManager.volume.value).toEqual(5)

      await this.user.click(screen.getByTitle('Mute'))
      expect(html()).toMatchSnapshot()
      expect(volumeManager.volume.value).toEqual(0)

      await this.user.click(screen.getByTitle('Unmute'))
      expect(html()).toMatchSnapshot()
      expect(volumeManager.volume.value).toEqual(5)
    })

    it('sets and broadcasts volume', async () => {
      const setMock = this.mock(volumeManager, 'set')
      const broadCastMock = this.mock(socketService, 'broadcast')
      this.render(Volume)

      await fireEvent.update(screen.getByRole('slider'), '4.2')
      await fireEvent.change(screen.getByRole('slider'))

      expect(setMock).toHaveBeenCalledWith(4.2)
      expect(broadCastMock).toHaveBeenCalledWith('SOCKET_VOLUME_CHANGED', 4.2)
    })
  }
}
