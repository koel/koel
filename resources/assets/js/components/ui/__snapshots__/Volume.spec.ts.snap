// Vitest Snapshot v1

exports[`mutes and unmutes 1`] = `<span id="volume" class="volume loud"><span role="button" tabindex="0" title="Unmute" style="display: none;"><br data-testid="icon" icon="[object Object]" fixed-width=""></span><span role="button" tabindex="0" title="Mute"><br data-testid="icon" icon="[object Object]" fixed-width=""></span><input id="volumeInput" class="plyr__volume" max="10" role="slider" step="0.1" title="Volume" type="range"></span>`;

exports[`mutes and unmutes 2`] = `<span id="volume" class="volume muted"><span role="button" tabindex="0" title="Unmute" style=""><br data-testid="icon" icon="[object Object]" fixed-width=""></span><span role="button" tabindex="0" title="Mute" style="display: none;"><br data-testid="icon" icon="[object Object]" fixed-width=""></span><input id="volumeInput" class="plyr__volume" max="10" role="slider" step="0.1" title="Volume" type="range"></span>`;

exports[`mutes and unmutes 3`] = `<span id="volume" class="volume loud"><span role="button" tabindex="0" title="Unmute" style="display: none;"><br data-testid="icon" icon="[object Object]" fixed-width=""></span><span role="button" tabindex="0" title="Mute" style=""><br data-testid="icon" icon="[object Object]" fixed-width=""></span><input id="volumeInput" class="plyr__volume" max="10" role="slider" step="0.1" title="Volume" type="range"></span>`;
